core/src/main/scala/org/apache/spark/TaskContextImpl.scala

```diff
@@ -57,57 +68,75 @@ private[spark] class TaskContextImpl(
   // Whether the task has failed.
   @volatile private var failed: Boolean = false
```

Comment:
If drop the volatility then we need to make `isCompleted` synchronized as well; to ensure safe publication.